indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java

```diff
@@ -0,0 +1,266 @@
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = pair.rhs;
+    final List<String> dimensions = IngestSegmentFirehoseFactory.getUniqueDimensions(
+        timelineSegments,
+        new NoopInputRowParser(null)
+    );
+    final List<String> metrics = IngestSegmentFirehoseFactory.getUniqueMetrics(timelineSegments);
+    return new IndexIngestionSpec(
+        createDataSchema(dataSource, interval, indexIO, jsonMapper, timelineSegments, segmentFileMap),
+        new IndexIOConfig(
+            new IngestSegmentFirehoseFactory(dataSource, interval, null, dimensions, metrics, injector, indexIO),
+            false
+        ),
+        tuningConfig
+    );
+  }
+
+  private static Pair<Map<DataSegment, File>, List<TimelineObjectHolder<String, DataSegment>>> prepareSegments(
+      TaskToolbox toolbox,
+      String dataSource,
+      Interval interval
+  ) throws IOException, SegmentLoadingException
+  {
+    final List<DataSegment> usedSegments = toolbox
+        .getTaskActionClient()
+        .submit(new SegmentListUsedAction(dataSource, interval, null));
+    final Map<DataSegment, File> segmentFileMap = toolbox.fetchSegments(usedSegments);
+    final List<TimelineObjectHolder<String, DataSegment>> timelineSegments = CompactionTaskUtils.toTimelineSegments(
+        usedSegments,
+        interval
+    );
+    return Pair.of(segmentFileMap, timelineSegments);
+  }
+
+  private static DataSchema createDataSchema(
+      String dataSource,
+      Interval interval,
+      IndexIO indexIO,
+      ObjectMapper jsonMapper,
+      List<TimelineObjectHolder<String, DataSegment>> timelineSegments,
+      Map<DataSegment, File> segmentFileMap
+  )
+      throws IOException, SegmentLoadingException
+  {
+    // find metadata for interval
+    final List<QueryableIndex> segments = CompactionTaskUtils.loadSegments(timelineSegments, segmentFileMap, indexIO);
+    final Granularity expectedGranularity = segments.get(0).getMetadata().getQueryGranularity();
+    final boolean expectedRollup = segments.get(0).getMetadata().isRollup();
+
+    // check metadata
+    Preconditions.checkState(
+        segments.stream().allMatch(index -> index.getMetadata().getQueryGranularity().equals(expectedGranularity))
+    );
+    Preconditions.checkState(
+        segments.stream().allMatch(index -> index.getMetadata().isRollup() == expectedRollup)
+    );
+
+    // find merged aggregators
+    final List<AggregatorFactory[]> aggregatorFactories = segments.stream()
+                                                                  .map(index -> index.getMetadata().getAggregators())
+                                                                  .collect(Collectors.toList());
+    final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(aggregatorFactories);
+    // conver to combining aggregators
+    final AggregatorFactory[] combiningAggregators;
+    if (mergedAggregators == null) {
+      combiningAggregators = null;
```

Comment:
Probably need to thrown an exception here. I think if we actually go through with a dataSchema with null aggregators, it will just drop all the metrics while compacting, which seems like a bad idea.

Reply:
Good point. I added a check.